src/Microsoft.OData.Client/DeepInsertSaveResult.cs

```diff
@@ -0,0 +1,303 @@
+            {
+                InvalidOperationException exception = WebUtil.GetHttpWebResponse(ex, ref this.batchResponseMessage);
+
+                throw exception;
+            }
+        }
+
+        /// <summary>
+        /// This method processes all the changed descriptors in the entity tracker.
+        /// It loops through all the descriptors and creates relationships between the descriptors if any.
+        /// </summary>
+        /// <typeparam name="T">The type of the top-level resource we want to do a deep insert.</typeparam>
+        /// <param name="descriptors">The list of descriptors in the entity tracker.</param>
+        /// <param name="isRootObject">True if the <paramref name="objects"/> are root or top-level objects, i.e., they were passed directly by the client, otherwise false.</param>
+        /// <param name="objects">Objects of the top-level type we want to perform a deep insert on.</param>
+        /// <exception cref="InvalidOperationException"></exception>
+        internal void BuildDescriptorGraph<T>(IEnumerable<Descriptor> descriptors, bool isRootObject, params T[] objects)
+        {
+            foreach (object obj in objects)
+            {
+                EntityDescriptor topLevelDescriptor = this.RequestInfo.EntityTracker.GetEntityDescriptor(obj);
+
+                // If we do not find an EntityDescriptor in the entitytracker
+                // for any of the provided objects then we throw an exception. 
+                if (topLevelDescriptor == null)
+                {
+                    throw Error.InvalidOperation(Strings.Context_EntityNotContained);
+                }
+
+                // If it's a top-level object sent by the user.
+                if (isRootObject)
+                {
+                    // Validate that we can only have one top level entity in Deep Insert.
+                    if (bulkUpdateGraph.TopLevelDescriptors.Count > 0)
+                    {
+                        throw Error.InvalidOperation(Strings.Context_DeepInsertOneTopLevelEntity);
+                    }
+
+                    bulkUpdateGraph.AddTopLevelDescriptor(topLevelDescriptor);
+                }
+
+                foreach (Descriptor descriptor in descriptors)
+                {
+                    if (descriptor is EntityDescriptor entityDescriptor)
+                    {
+                        if (entityDescriptor.ParentEntity != null &&
+                            entityDescriptor.ParentEntity.Equals(topLevelDescriptor.Entity) &&
+                            !bulkUpdateGraph.Contains(entityDescriptor))
+                        {
+                            // We don't support delete and update in deep insert.
+                            if (entityDescriptor.State == EntityStates.Deleted || entityDescriptor.State == EntityStates.Modified)
+                            {
+                                throw Error.InvalidOperation(Strings.Context_DeepInsertDeletedOrModified);

```

Comment:
Actually not sure.

Reply:
I think this can be documented so that users are aware that a deep insert is only used for create operations and linking of existing resources to newly created resources. If a user has this kind of information then I think supporting the second option of not throwing an exception then the user being required to call `SaveChanges` to save the other operations that were not taken care of by the `DeepInsert.`  My thoughts though.

Reply:
@KenitoInc  what was agreed on here? Are you throwing exceptions?